src/main/scala/com/github/pathikrit/dijon/package.scala

```diff
@@ -135,71 +131,93 @@ package object dijon {
-      case 'n' => in.readNullOrError(default, "expected `null` value")
-      case '"' =>
+    override def decodeValue(in: JsonReader, default: SomeJson): SomeJson = decode(in, maxParsingDepth)
+
+    override def encodeValue(x: SomeJson, out: JsonWriter): Unit = encode(x, out, maxSerializationDepth)
+
+    override val nullValue: SomeJson = None
+
+    private[this] def decode(in: JsonReader, depth: Int): SomeJson = {
+      val b = in.nextToken()
+      if (b == 'n') in.readNullOrError(None, "expected `null` value")
+      else if (b == '"') {
         in.rollbackToken()
         in.readString(null)
-      case 'f' | 't' =>
+      } else if (b == 't' || b == 'f') {
         in.rollbackToken()
         in.readBoolean()
-      case '0' | '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' | '-' =>
+      } else if ((b >= '0' && b <= '9') || b == '-') {
         in.rollbackToken()
         val d = in.readDouble()
         val i = d.toInt
         if (i.toDouble == d) i
         else d
-      case '[' =>
-        val arr = new mutable.ArrayBuffer[SomeJson]
+      } else if (b == '[') {
+        if (depth <= 0) in.decodeError("depth limit exceeded")
+        val arr = new mutable.ArrayBuffer[SomeJson](initArrayCapacity)
         if (!in.isNextToken(']')) {
           in.rollbackToken()
-          do arr += decodeValue(in, default)
+          val dp = depth - 1
+          do arr += decode(in, dp)
           while (in.isNextToken(','))
           if (!in.isCurrentToken(']')) in.arrayEndOrCommaError()
         }
         arr
-      case '{' =>
-        val obj = new util.LinkedHashMap[String, SomeJson]
+      } else if (b == '{') {
+        if (depth <= 0) in.decodeError("depth limit exceeded")
+        val obj = new util.LinkedHashMap[String, SomeJson](initMapCapacity)
         if (!in.isNextToken('}')) {
           in.rollbackToken()
-          do obj.put(in.readKeyAsString(), decodeValue(in, default))
+          val dp = depth - 1
+          do obj.put(in.readKeyAsString(), decode(in, dp))
           while (in.isNextToken(','))
           if (!in.isCurrentToken('}')) in.objectEndOrCommaError()
         }
         obj.asScala
-      case _ => in.decodeError("expected JSON value")
+      } else in.decodeError("expected JSON value")
     }
 
-    override def encodeValue(x: SomeJson, out: JsonWriter): Unit = x.underlying match {
+    private[this] def encode(x: SomeJson, out: JsonWriter, depth: Int): Unit = x.underlying match {
       case None => out.writeNull()
       case str: String => out.writeVal(str)
       case b: Boolean => out.writeVal(b)
       case i: Int => out.writeVal(i)
       case d: Double => out.writeVal(d)
       case arr: JsonArray =>
+        if (depth <= 0) out.encodeError("depth limit exceeded")
         out.writeArrayStart()
-        arr.foreach(v => encodeValue(v, out))
+        val it = arr.iterator
+        val dp = depth - 1
+        while (it.hasNext) encode(it.next(), out, dp)
```

Comment:
Have you considered using a `var i: Int` and indexing into the array? GraalVM EE can probably remove the iterator allocation, but I'd expect it to show up under the profiler in OpenJDK.